open_lm/main.py

```diff
@@ -707,34 +685,35 @@ def main(args):
         if args.distributed:
             dist.barrier()
 
+        done_training = global_step >= total_steps
         steps_done_epoch = global_step - prev_step
         samples_seen = samples_seen + steps_done_epoch * args.batch_size * args.world_size
 
         if not success:
             logging.info("Training exiting due to NaN value")
             break
 
-        completed_epoch = epoch + 1
+        epoch = epoch + 1
         evaluation_loss = -1
-        if "val" in data and (completed_epoch % args.val_frequency == 0 or completed_epoch == args.epochs):
+        if "val" in data and (epoch % args.val_frequency == 0 or done_training):
             # validate based on frequency and always validate the last checkpoint
-            evaluation_loss = evaluate(model, data, completed_epoch, args, writer)["loss"]
+            evaluation_loss = evaluate(model, data, epoch, args, writer)["loss"]
 
         # Saving checkpoints.
         save_checkpoint(
             args,
             model,
             optimizer,
             scaler,
-            completed_epoch,
+            epoch,
             evaluation_loss,
             step=global_step,
-            next_chunk=next_chunk if args.dataset_manifest is not None else None,
+            next_shard_per_source=next_shard_per_source if args.dataset_manifest is not None else None,
             samples_seen=samples_seen if args.dataset_manifest is not None else None,
         )
 
-        if final_epoch:
-            logging.info("Ending training due to data exhaustion.")
+        if is_master(args) and done_training:
+            logging.info("Model has seen the desired number of tokens. Ending training.")
             break
```

Comment:
Should we break in all the workers, not just rank 0?

Reply:
Good catch!